safe-global/safe-wallet-web
# feat: add push notifications

## What it solves

Implements push notifications for transaction events.

## How this PR fixes it

New "Notifications" settings sections have been added per-Safes/all added Safes, as well as a feature banner and associated tracking has been added. Subscribed categories dispatch push notifccations when the Safe UI is either out of focus or closed (but the browser is open).

The per-Safe "Notifications" section allows for notification subscription to all notification types, requiring a confirmatory signature as "Confirmation requests" are included. The global section is similar, subscribing to user selected (added) Safes with n number of signatures per selected chain.

Notifications are grouped into three categories: "Incoming transactions", "Outgoing transactions" and "Confirmation requests" according to [our Transaction Service webhooks](https://github.com/safe-global/safe-transaction-service/blob/master/safe_transaction_service/notifications/tasks.py).

- Incoming transactions
  - `INCOMING_ETHER`
  - `INCOMING_TOKEN`
- Outgoing transactions
  - `MODULE_TRANSACTION`
  - `EXECUTED_MULTISIG_TRANSACTION`
- Confirmation requests
  - `CONFIRMATION_REQUEST`

A banner is shown once per chain that has added Safes, subscribing to notifications for all added Safes on that chain.

User interaction within the UI is tracked directly. ~Due to technical limitations, direct notification interaction events (opening/closing) are cached and tracked when the UI is again opened.~ Moved to a separate PR https://github.com/safe-global/safe-wallet-web/pull/2500.

## How to test it

⚠️ Due to the complexity of this feature, I would advise explaining the feature on a call and maybe even testing it together. ⚠️

1. Registering for notifications should succeed, showing a notification and then displayind them accordingly when the Safe UI is not in focus or is closed (but the browser is open).
2. Unregistering a Safe should only unregister itself if there are others subscribed to on that chain or the entire device if it is the last Safe subscribed to on that chain.
3. Tracking of [user interaction](https://github.com/safe-global/safe-wallet-web/blob/ceaacff11584fbcebecfbd2f449c6a72b073825c/src/services/analytics/events/push-notifications.ts#L14-L95) should happen immediately.
4. ~Tacking of [notification iteraction](https://github.com/safe-global/safe-wallet-web/blob/ceaacff11584fbcebecfbd2f449c6a72b073825c/src/services/analytics/events/push-notifications.ts#L5-L13) should happen when the UI is reopened.~ (Moved to a separate PR https://github.com/safe-global/safe-wallet-web/pull/2500).
5. The MacOS info banner should only show on Mac devices.

If you want to reset your current registration, you need to:

- Reset browser notification permissions
- Unregister the `sw.js` service worker
- Delete all `firebase-*`/`notification-*` IndexedDB databases
![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/d65f4885-4b7b-474d-857e-d2680839caf2)

If you want to show the subscription banner again, you need to delete the `SAFE_v2__dismissPushNotifications` `localStorage` key.

## Screenshots

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/47e17311-e48b-43c9-82b5-87fd373a27c4)

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/58ef44f5-bccb-441c-a4fb-780a73dd8411)

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/34b0db2b-9882-47b0-8bec-097726d3a36e)

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/0d499dee-ec22-477c-862d-513b389f547f)

Please note that these are slightly dated but just demonstrate the notification. See the [designs](https://www.figma.com/file/ma5syWag1CO8yVbuzqkK0u/Push-notifications?mode=dev).

![notification](https://github.com/safe-global/safe-wallet-web/assets/20442784/2cb1f65a-58d3-41a5-89d1-817a19984fed)

## Checklist
